src/config/options.rs

```diff
@@ -105,11 +105,68 @@ macro_rules! configuration_option_enum{
 }
 
 configuration_option_enum! { NewlineStyle:
+    Auto, // Auto-detect based on the raw source input
     Windows, // \r\n
     Unix, // \n
     Native, // \r\n in Windows, \n on other platforms
 }
 
+impl NewlineStyle {
+    fn auto_detect(raw_input_text: &str) -> NewlineStyle {
+        if let Some(pos) = raw_input_text.find('\n') {
+            let pos = pos.saturating_sub(1);
+            if let Some('\r') = raw_input_text.chars().nth(pos) {
+                NewlineStyle::Windows
+            } else {
+                NewlineStyle::Unix
+            }
+        } else {
+            NewlineStyle::Native
+        }
+    }
+
+    fn native() -> NewlineStyle {
+        if cfg!(windows) {
+            NewlineStyle::Windows
+        } else {
+            NewlineStyle::Unix
+        }
+    }
+
+    /// Apply this newline style to the formatted text. When the style is set
+    /// to `Auto`, the `raw_input_text` is used to detect the existing line
+    /// endings.
+    ///
+    /// If the style is set to `Auto` and `raw_input_text` contains no
+    /// newlines, the `Native` style will be used.
+    pub(crate) fn apply(self, formatted_text: &mut String, raw_input_text: &str) {
+        use NewlineStyle::*;
+        let mut style = self;
+        if style == Auto {
+            style = Self::auto_detect(raw_input_text);
+        }
+        if style == Native {
+            style = Self::native();
+        }
+        match style {
+            Windows => {
+                let mut transformed = String::with_capacity(formatted_text.capacity());
```

Comment:
This seems like the capacity will never be enough? I think you should use `formatted_text.len() + formatted_text.match('\n').count()` (or some similar heuristic). It is better to over-estimate the capacity than to slightly under-estimate, under-estimation leads to a guaranteed re-allocation and copy of nearly the whole string.

Reply:
I'm not sure where the `\r` is coming from - maybe comments or string literals? I'm pretty sure Rustfmt only inserts `\n` when formatting.

Reply:
So, I think we should still over-allocate the `String`s - there is very little harm in over-allocating, but under-allocating could cause non-trivial slow down.